### PR TITLE
first changes to try get legacy provider up and running within silex 2.0

### PR DESCRIPTION
**Pull request by dominikzogg**
Hello @stof 

I tried to create a legacy provider support as we discus about today. Its in a early stage and not working correctly at the moment.

One problem for example would be provider which tried to create new \Pimple instances.

https://github.com/silexphp/Silex/blob/1.3/src/Silex/Provider/DoctrineServiceProvider.php#L65

Regards @dominikzogg

### CODE REVIEW

**Comment by dominikzogg** [comment]
@stof ok its working now, which means the old monolog provider is usable within silex 2. but i had to change some code withinin its tests, cause log level changed, and override some services needs to be done in a pimple 3 way.

File: src/Silex/Application.php [hunk 1]
```diff
@@ -47,6 +49,7 @@ class Application extends Container implements HttpKernelInterface, TerminableIn
     const LATE_EVENT = -512;
 
     protected $providers = array();
+    protected $legacyProvider = false;
```
Comment (stof) [inline src/Silex/Application.php]: this should be private, to avoid letting child classes messing with it

File: src/Silex/Application.php [hunk 2]
```diff
@@ -93,6 +96,30 @@ public function register(ServiceProviderInterface $provider, array $values = arr
...
+        $this->providers[] = $provider;
+        $this->legacyProvider = true;
+
+        $provider->register($this);
+
+        foreach ($values as $key => $value) {
+            $this[$key] = $value;
+        }
+
+        $this->legacyProvider = false;
```
Comment (stof) [inline src/Silex/Application.php]: you need to reset this also when an exception happens during the registration

File: src/Silex/Application.php [hunk 3]
```diff
@@ -107,6 +134,16 @@ public function boot()
         $this->booted = true;
 
         foreach ($this->providers as $provider) {
+            if ($provider instanceof LegacyServiceProviderInterface) {
+                $this->legacyProvider = true;
+
+                $provider->boot($this);
+
+                $this->legacyProvider = false;
```
Comment (stof) [inline src/Silex/Application.php]: same issue here

File: src/Silex/Application.php [hunk 4]
```diff
@@ -503,4 +540,65 @@ public function terminate(Request $request, Response $response)
...
+            parent::offsetSet($id, parent::factory($value));
+
+            return;
+        }
+
+        parent::offsetSet($id, $value);
+    }
+
+    /**
+     * Returns a closure that stores the result of the given service definition
```
Comment (stof) [inline src/Silex/Application.php]: wrong description

File: src/Silex/ServiceProviderInterface.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+namespace Silex;
+
+/**
+ * Interface that all Silex service providers must implement.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ * 
+ * @deprecated
```
Comment (stof) [inline src/Silex/ServiceProviderInterface.php]: missing deprecation message

File: src/Silex/Legacy/SharedService.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Silex\Legacy;
+
+class SharedService
```
Comment (stof) [inline src/Silex/Legacy/SharedService.php]: the class should be marked as internal

**Comment by stof** [comment]
Is BC provided for ``extend`` too ?

File: src/Silex/Legacy/SharedService.php (hunk 6 above)
Comment (stof) [inline src/Silex/Legacy/SharedService.php]: and you may want to make it invokable to be more BC

**Comment by GromNaN** [comment]
It might be a stupid question, but why not upgrade existing service providers for Silex2 / Pimple3 ?

All service providers of the core of Silex have been upgraded quite easily. What are you trying to use that requires so much effort that you find it easier to modify both Pimple and Silex than that code ?

**Comment by dominikzogg** [comment]
@GromNaN the goal was to provide an easy to use solutions for old providers. At the moment its only a proof of concept.

**Comment by dominikzogg** [comment]
@stof yeah, bc of extend should be given

**Comment by dominikzogg** [comment]
@stof why do think that SharedService needs to be invokable? its only used as helper within Application class

**Comment by dominikzogg** [comment]
@fabpot the problem, that i see with such a solution is, that every legacy provider which creates its own pimple instances as for example dbal provider does, would not work. Which means there would be pimple changes too, and it get very ugly.

https://github.com/silexphp/Silex/blob/1.3/src/Silex/Provider/DoctrineServiceProvider.php#L65

**Comment by fabpot** [comment]
@dominikzogg Indeed :( Perhaps you were right when we discussed that topic; it's maybe not worth it.

**Comment by dominikzogg** [comment]
@fabpot i close this PR and the one against pimple

**Pull request closed without merging**